### PR TITLE
Development v6

### LINKED ISSUE #2424: Web Interface locks up after being up open for a long time.
### Versions

- Pi-hole:  5.12.2
- AdminLTE:  5.15.1
- FTL:  5.18.1

### Platform

- OS and version:  Win11 

- Platform:  Raspberry Pi

### Expected behavior

_After having the Web UI open for a while (A few days?) I click on disable blocking -> 5 min. I'm hoping that the Blocking gets disabled._

### Actual behavior / bug

_What actually happens (After the UI is open a while). Clicking on Disable Blocking ->For 5 Minutes. The spinner shows, but it doesn't spin. I need to actually refresh the page to actually be able to disable blocking_

### Steps to reproduce

1. Open Web Interface and leave open for a few hours (1 day?)
2. Click on Disable Blocking -> For 5 Minutes
4. See error: Spinner shows but doesn't spin. Doesn't actually disable blocking.

### Additional context

Super low priority, bit this has been present for quite a few versions. Ill confirm that it shows up in the newest version.

### LINKED ISSUE #2606: Adding/Deleting multiple Custom DNS Records in a short amount of time triggers systemd service StartLimitBurst leading to disable pihole-FTL service
### Versions

- Pi-hole:  v5.17.1
- AdminLTE:  v5.20.1
- FTL:  v5.23

### Platform

- OS and version:  Ubuntu 22.04.2 LTS (5.15.0-73-generic)

- Platform:  Proxmox-VM

### Expected behavior
Adding or Deleting custom DNS records through webadmin GUI or respectively directly through POST requests to the API endpoint at `/admin/scripts/pi-hole/php/customdns.php` causes pihole-FTL to just be `reloaded`.

### Actual behavior / bug
repeatedly adding or deleting DNS records through the web interface (`/admin/scripts/pi-hole/php/customdns.php`) causes pihole-FTL to be **restarted** (not _reloaded_) with each and every DNS record added. This leads to systemd hitting the `StartLimitBurst=5` configured in `/etc/systemd/system/pihole-FTL.service` leading to consecutive restarts of the service to fail.

systemctl status pihole-FTL:
```
× pihole-FTL.service - Pi-hole FTL
     Loaded: loaded (/etc/systemd/system/pihole-FTL.service; enabled; vendor preset: enabled)
     Active: failed (Result: start-limit-hit) since Fri 2023-06-09 19:26:53 UTC; 17s ago
    Process: 3209 ExecStartPre=/opt/pihole/pihole-FTL-prestart.sh (code=exited, status=0/SUCCESS)
    Process: 3222 ExecStart=/usr/bin/pihole-FTL -f (code=exited, status=0/SUCCESS)
    Process: 3245 ExecStopPost=/opt/pihole/pihole-FTL-poststop.sh (code=exited, status=0/SUCCESS)
   Main PID: 3222 (code=exited, status=0/SUCCESS)
        CPU: 165ms

Jun 09 19:26:53 pihole01 pihole-FTL[3222]: [2023-06-09 19:26:52.992 3222M] Resizing "FTL-queries" from 4587520 to (86016 * 56) == 4816896 (/dev/shm: 5.1MB used, 4.2GB total, FTL uses 5.1MB)
Jun 09 19:26:53 pihole01 pihole-FTL[3222]: [2023-06-09 19:26:52.997 3222M] Resizing "FTL-queries" from 4816896 to (90112 * 56) == 5046272 (/dev/shm: 5.4MB used, 4.2GB total, FTL uses 5.3MB)
Jun 09 19:26:53 pihole01 pihole-FTL[3222]: [2023-06-09 19:26:52.001 3222M] Resizing "FTL-queries" from 5046272 to (94208 * 56) == 5275648 (/dev/shm: 5.6MB used, 4.2GB total, FTL uses 5.6MB)
Jun 09 19:26:53 pihole01 pihole-FTL[3222]: [2023-06-09 19:26:53.007 3222M] Resizing "FTL-queries" from 5275648 to (98304 * 56) == 5505024 (/dev/shm: 5.8MB used, 4.2GB total, FTL u
Jun 09 19:26:53 pihole01 systemd[1]: Stopping Pi-hole FTL...
Jun 09 19:26:53 pihole01 systemd[1]: pihole-FTL.service: Deactivated successfully.
Jun 09 19:26:53 pihole01 systemd[1]: Stopped Pi-hole FTL.
Jun 09 19:26:53 pihole01 systemd[1]: pihole-FTL.service: Start request repeated too quickly.
Jun 09 19:26:53 pihole01 systemd[1]: pihole-FTL.service: Failed with result 'start-limit-hit'.
Jun 09 19:26:53 pihole01 systemd[1]: Failed to start Pi-hole FTL.
```

#### Shell
1. run this shell script
```
/usr/local/bin/pihole -a addcustomdns 127.0.0.1 test1.local
/usr/local/bin/pihole -a addcustomdns 127.0.0.1 test2.local
/usr/local/bin/pihole -a addcustomdns 127.0.0.1 test3.local
/usr/local/bin/pihole -a addcustomdns 127.0.0.1 test4.local
/usr/local/bin/pihole -a addcustomdns 127.0.0.1 test5.local
/usr/local/bin/pihole -a addcustomdns 127.0.0.1 test6.local
```
2. systemctl status pihole-FTL will now be `killed` and fail to start with `Failed with result 'start-limit-hit'`

#### Webadmin
1. Add or delete 6 DNS records in under 60 seconds.
2. systemctl status pihole-FTL will now be `killed` and fail to start with `Failed with result 'start-limit-hit'`

## Debug Token
I don't think this should be necessary here.

### Additional context
I am using the pihole provider from [kubernetes-sigs/external-dns](https://github.com/kubernetes-sigs/external-dns/tree/master/provider/pihole) in combination with [borchero/switchboard](https://github.com/borchero/switchboard/) to automatically add DNS records for services in my cluster when they get exposed. This worked fine while setting everything up.

However, after exposing more services for external-dns to manage DNS for, pihole-FTL started to behave oddly/refused to start.
The way [kubernetes-sigs/external-dns](https://github.com/kubernetes-sigs/external-dns/tree/master/provider/pihole) adds the dns records, is by just simply sending POST requests to the same endpoint the webadmin GUI uses (`/admin/scripts/pi-hole/php/customdns.php`). Due to this API endpoint not offering any `update` functionality and also the fact that no `TXT` records are possible, this results in a little bit of a "spammy" behaviour from external-dns. As far as i understand the implementation of the pihole provider on external-dns, there are a couple workarounds in place, due to the API limitations of pihole-FTL. There might also be another bug in the pihole provider for external-dns, but I have not yet spent the time to dig into that side any deeper and I believe that even if this was the case, it would not change validity of this bug report.

For a quick and dirty workaround I adjusted `StartLimitBurst` in `/etc/systemd/system/pihole-FTL.service`. 

When digging through the source code of pihole I noticed a couple things and I am not exactly sure where this should be fixed.

1. https://github.com/pi-hole/AdminLTE/blob/3a11976ee8ecc50e2dd9efd76caad1ad41894dd5/scripts/pi-hole/php/func.php#L223
empty string as default value for the `reload` argument.

2. https://github.com/pi-hole/AdminLTE/blob/3a11976ee8ecc50e2dd9efd76caad1ad41894dd5/scripts/pi-hole/php/func.php#L280
effectively results in `pihole -a addcustomdns 127.0.0.1 test1.local` (notice neither `true` nor `false` are part of the argument for restart)

3. https://github.com/pi-hole/pi-hole/blob/6a45c6a8e027e1ac30d4556a88f31684bc80ccf1/pihole#L579
`pihole` shell script defers to `AddCustomDNSAddress` in `webpage.sh`

4. https://github.com/pi-hole/pi-hole/blob/6a45c6a8e027e1ac30d4556a88f31684bc80ccf1/advanced/Scripts/webpage.sh#L719-L743
RestartDNS command will be issued due to missing empty reload argument

5. https://github.com/pi-hole/AdminLTE/blob/3a11976ee8ecc50e2dd9efd76caad1ad41894dd5/scripts/pi-hole/php/func.php#L284
RestartDNS triggered again. Should this be `restartdns reload-lists` instead of just `restartdns` here?

#### Conclusion
The issue here seems to affect both the AdminLTE as well as the pihole repository, so I am not entirely sure how the best or preferred way of fixing this would look like.

### PR DESCRIPTION
**Pull request by PromoFaux**
Step one of pre-release, merge `development-v6` into `development` branch

### CODE REVIEW

**Comment by PromoFaux** [comment]
lol @yubiuser - still conflicts 😬

**Comment by yubiuser** [comment]
Ahh, but why?

**Comment by PromoFaux** [comment]
@DL6ER / @rdwebdesign do we want to do something about the CodeQL report or just dismiss it?

File: scripts/pi-hole/js/login.js [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    // If specified: Send to requested page
+    var GETDict = getParams();
+    if ("target" in GETDict) {
+      // URL-decode target
+      target = GETDict.target;
+    }
+  }
+
+  // Redirect to target
+  window.location.replace(target);
```
[the remark this answers is not in the document]
Comment (rdwebdesign) [reply scripts/pi-hole/js/login.js:21]: This is one confirmed.

When you try to access a page BEFORE login, you will be redirected to the login page, but after that, the login page will redirect to the original destination.

**Example:**
Trying to access `http://pi.hole/admin/settings` without authentication will redirect to the login form and after that you will be redirect to the Settings page.

**The issue:**
If you try `http://pi.hole/admin/login?target=https://example.com`, after a successful login you will be redirected to `https://example.com`.

The v5 behavior is similar, but the URL is passed internally via PHP session, so it's not possible to inject an URL, but in v6 the destination URL is passed on the URL.

[the remark this answers is not in the document]
Comment (PromoFaux) [reply scripts/pi-hole/js/login.js:21]: I guess that should probably be mitigated then before release!

**Comment by DL6ER** [comment]
I will open a PRs to remove the redirection feature

File: scripts/pi-hole/js/login.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (DL6ER) [reply scripts/pi-hole/js/login.js:21]: Fixed

**Review by DL6ER — APPROVED** [review]

**Pull request merged**